proposals/lip-0051.md

```diff
@@ -503,240 +637,1915 @@ crossChainTransferMessageParams = {
+            emitFailedCCMForwardEvent(ccm.sendingChainID, ccm.receivingChainID, messageFeeTokenID, ccm.fee, INSUFFICIENT_ESCROW_BALANCE) 
+            raise Exception("Insufficient balance in the sending chain for the transfer.")
+
+        # Transfer the amount from escrow to escrow
+        escrowAmount(ccm.sendingChainID, ccm.params.tokenID) -= ccm.params.amount
+        escrowAmount(ccm.receivingChainID, ccm.params.tokenID) += ccm.params.amount
+
+        emitEvent(
+            module=MODULE_NAME_TOKEN,
+            name=EVENT_NAME_BEFORE_CCM_FORWARDING,
+            data={
+                "sendingChainID": ccm.sendingChainID, 
+                "receivingChainID":ccm.receivingChainID,
+                "messageFeeTokenID": messageFeeTokenID, 
+                "messageFee": ccm.fee, 
+                "result": RESULT_SUCCESSFUL 
+            },
+            topics=[ccm.sendingChainID, ccm.receivingChainID]
+        )
+
+def emitFailedCCMForwardEvent(
+    sendingChainID: ChainID,
+    receivingChainID: ChainID,
+    messageFeeTokenID: TokenID,
+    messageFee: uint64,
+    result: uint32
+) -> None:
+
+    emitPersistentEvent(
+        module=MODULE_NAME_TOKEN,
+        name=EVENT_NAME_BEFORE_CCM_FORWARDING,
+        data={
+            "sendingChainID": sendingChainID, 
+            "receivingChainID": receivingChainID,
+            "messageFeeTokenID": messageFeeTokenID, 
+            "messageFee": messageFee, 
+            "result": result 
+        },
+        topics=[ccm.sendingChainID, ccm.receivingChainID]
+    )
+```
+
+#### verifyCrossChainMessage
+
+```python
+def verifyCrossChainMessage(ccu: Transaction, ccm: CCM) -> None:
+    # This is TOKEN_ID_LSK for channels with the mainchain.
+    feeTokenID = Interoperability.getMessageFeeTokenID(ccm.sendingChainID)
+    # If the chain is the fee token native chain, un-escrow and assign fee to relayer.
+    if getChainID(feeTokenID) == OWN_CHAIN_ID:
+        if escrowAmount(ccm.sendingChainID, feeTokenID) < ccm.fee:
+            raise Exception("Insufficient escrow amount.")
+``` 
+
+#### recover
+
+```python
+def recover(
+    terminatedChainID: ChainID, 
+    module: Module, 
+    substorePrefix: bytes, 
+    storeKey: bytes, 
+    storeValue: bytes
+) -> None:
+    if (
+        substorePrefix != SUBSTORE_PREFIX_USER
+        or len(storeKey) != ADDRESS_LENGTH + TOKEN_ID_LENGTH
+        or storeValue cannot be deserialized using userStoreSchema
+    ):
+
+        emitFailedRecoverEvent(terminatedChainID, storeKey[:ADDRESS_LENGTH], EMPTY_BYTES, 0, RECOVER_FAIL_INVALID_INPUTS)
+        raise Exception("Invalid arguments.")
+
+    address = storeKey[:ADDRESS_LENGTH]
+    tokenID = storeKey[ADDRESS_LENGTH:ADDRESS_LENGTH + TOKEN_ID_LENGTH]
+    account = decode(schema=userStoreSchema, object=storeValue)
+    totalAmount = sum of availableBalance and all locked amounts of account
+
+    if (
+        getChainID(tokenID) != OWN_CHAIN_ID
```

Comment:
In this case we should emit an event with a different error code and raise a different Exception (insufficient escrow is not really correct)